core/services/relay/evm/chain_writer.go

```diff
@@ -156,7 +169,38 @@ func (w *chainWriter) GetTransactionStatus(ctx context.Context, transactionID uu
 }
 
 func (w *chainWriter) GetFeeComponents(ctx context.Context) (*commontypes.ChainFeeComponents, error) {
-	return nil, fmt.Errorf("not implemented")
+	if w.ge.estimator == nil {
+		return nil, fmt.Errorf("not implemented")
+	}
+
+	gasPriceWei, _, err := w.ge.estimator.GetFee(ctx, nil, 0, assets.NewWei(w.ge.maxGasPrice))
+	if err != nil {
+		return nil, err
+	}
+	// Use legacy if no dynamic is available.
+	gasPrice := gasPriceWei.Legacy.ToInt()
+	if gasPriceWei.DynamicFeeCap != nil {
+		gasPrice = gasPriceWei.DynamicFeeCap.ToInt()
+	}
+	if gasPrice == nil {
+		return nil, fmt.Errorf("missing gas price %+v", gasPriceWei)
+	}
+	l1Oracle := w.ge.estimator.L1Oracle()
+	if l1Oracle == nil {
+		return &commontypes.ChainFeeComponents{
+			ExecutionFee:        *gasPrice,
+			DataAvailabilityFee: *big.NewInt(0),
+		}, nil
+	}
+	l1OracleFee, err := l1Oracle.GasPrice(ctx)
```

Comment:
I wonder what kind of errors are possible here. Are there situations where we'd want to return both the executionFee and the l1Oracle RPC error?

Reply:
Hm, are you suggesting we don't return an error if the gasPrice errors out and wait to see if the L1Oracle is erroring out too, returning both errors in that case?

Reply:
Hm, I'm not sure actually. I think we would want to return nil and an error because the L1Oracle fee is required for estimating gas. I suppose it doesn't hurt to return the execution fee though

Reply:
The usual practice for returning a tuple that follows the `(value, error)` is for one and only one to be non-nil. I'd return `nil` here.